src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+
+At first I tried to use `requestDevice()` in the implementation of `get_device_list` handler to make usage easier. However, showing a permission prompt with a list of connected devices is considered a sensitive operation, and it must be triggered by user interaction (like a button click on a page), otherwise it always returns a rejected promise. libusb applications might often want to list the connected devices upon application start-up, so using `requestDevice()` was not an option.
+
+Instead, I had to leave invocation of `navigator.usb.requestDevice()` to the end developer, and only expose the already approved devices from `navigator.usb.getDevices()`:
+
+```cpp
+// Store the global `navigator.usb` once upon initialisation.
+thread_local const val web_usb = val::global("navigator")["usb"];
+
+int em_get_device_list(libusb_context *ctx, discovered_devs **devs) {
+  // C++ equivalent of `await navigator.usb.getDevices()`.
+  // Note: at this point we must already have some devices exposed -
+  // caller must have called `await navigator.usb.requestDevice(...)`
+  // in response to user interaction before going to LibUSB.
+  // Otherwise this list will be empty.
+  auto result = promise_result::await(web_usb.call<val>("getDevices"));
+  if (result.error) {
+    return result.error;
+  }
+  auto &web_usb_devices = result.value;
+  // Iterate over the exposed devices.
+  uint8_t devices_num = web_usb_devices["length"].as<uint8_t>();
+  for (uint8_t i = 0; i < devices_num; i++) {
+    auto web_usb_device = web_usb_devices[i];
+    // …
+    *devs = discovered_devs_append(*devs, dev);
+  }
+  return LIBUSB_SUCCESS;
+}
+```
+
+Most of the backend code uses `val` and `promise_result` in a similar way as already shown above. There were few more interesting hacks I've had to do to convert properties exposed by WebUSB to those expected by libusb, both in this method and in others handling data transfer, but those implementation details are less important for the purposes of this article. Make sure to check the code and the comments if you're interested.
+
+## Porting event loops to the web
+
+One more piece of `libusb` that I do want to stop on is event handling. As described in the previous article, most APIs in system languages like C are synchronous, and event handling is no exception. It's usually implemented via an infinite loop that "polls" (tries to read data or blocks execution until some data is available) from a set of external I/O sources, and, when at least one of those responds, passes that as an event to the corresponding handler. Once the handler has finished its job, the control returns to the loop, and it pauses for another poll.
```

Comment:
```suggestion
One more piece of `libusb` that I want to discuss is event handling. As described in the previous article, most APIs in system languages like C are synchronous, and event handling is no exception. It's usually implemented via an infinite loop that "polls" (tries to read data or blocks execution until some data is available) from a set of external I/O sources, and, when at least one of those responds, passes that as an event to the corresponding handler. Once the handler is finished, the control returns to the loop, and it pauses for another poll.
```

Reply:
>most APIs in system languages like C are synchronous

C is a low-level language, but I've never heard it described as a system language. Do you mean, "most system APIs used in C are synchronous"? Something elese?

Reply:
"System languages" is a fairly common term. https://en.wikipedia.org/wiki/System_programming_language